### PR TITLE
[clang][LoongArch] Add FreeBSD targets

### PR DESCRIPTION
**Pull request by hitmoon**
Add support for freebsd on loongarch

### CODE REVIEW

**Review by SixWeining — COMMENTED** [review]
Can FreeBSD be built with this PR?

File: clang/test/Driver/freebsd.c [hunk 1]
```diff
@@ -77,6 +77,15 @@
 // RUN:   | FileCheck --check-prefix=CHECK-RV64I-LD %s
 // CHECK-RV64I-LD: ld{{.*}}" {{.*}} "-m" "elf64lriscv"
 //
+// Check that LOONGARCH passes the correct linker emulation.
```
[the remark this answers is not in the document]
Comment (hitmoon) [reply clang/test/Driver/freebsd.c:80]: ok

**Comment by hitmoon** [comment]
> Can FreeBSD be built with this PR?

Yes,  tested by building kernel-toolchain

File: clang/test/Driver/freebsd.c
[the remark this answers is not in the document]
Comment (hitmoon) [reply clang/test/Driver/freebsd.c:1]: will add in next version

File: clang/test/Driver/freebsd.c [hunk 2]
```diff
@@ -77,6 +77,15 @@
...
+//
+// RUN: %clang --target=loongarch32-freebsd -### %s %s 2>&1 \
+// RUN:   | FileCheck --check-prefix=CHECK-LOONG32-LD %s
+// CHECK-LOONG32-LD: ld{{.*}}" {{.*}} "-m" "elf32loongarch"
+// RUN: %clang --target=loongarch64-freebsd -### %s %s 2>&1 \
+// RUN:   | FileCheck --check-prefix=CHECK-LOONG64-LD %s
+// CHECK-LOONG64-LD: ld{{.*}}" {{.*}} "-m" "elf64loongarch"
```
[the remark this answers is not in the document]
Comment (hitmoon) [reply clang/test/Driver/freebsd.c]: changed in new version

**Comment by SixWeining** [comment]
LGTM from the LoongArch side. It would be better a FreeBSD expert could have a look. But if there is no obvious objections I'd like to merge and mention it in LLVM20's release notes. Thanks! @hitmoon Looking forward to hear more news about the FreeBSD-LoongArch porting progress.

File: clang/lib/Driver/ToolChains/FreeBSD.cpp [hunk 3]
```diff
@@ -223,6 +231,12 @@ void freebsd::Linker::ConstructJob(Compilation &C, const JobAction &JA,
       CmdArgs.push_back("--no-relax");
   }
 
+  if (Triple.isLoongArch64()) {
+    CmdArgs.push_back("-X");
+    if (Args.hasArg(options::OPT_mno_relax))
+      CmdArgs.push_back("--no-relax");
```
Comment (xen0n) [inline clang/lib/Driver/ToolChains/FreeBSD.cpp:237]: A sentence or two describing the justification for the (hopefully temporary) disabling could be better.

**Comment by hitmoon** [comment]
> LGTM from the LoongArch side. It would be better a FreeBSD expert could have a look. But if there is no obvious objections I'd like to merge and mention it in LLVM20's release notes. Thanks! @hitmoon Looking forward to hear more news about the FreeBSD-LoongArch porting progress.

@SixWeining Thanks for the review !

File: clang/lib/Driver/ToolChains/FreeBSD.cpp (hunk 3 above)
Comment (hitmoon) [reply clang/lib/Driver/ToolChains/FreeBSD.cpp:237]: > A sentence or two describing the justification for the (hopefully temporary) disabling could be better.
here, I just construct the clang linker options from cmdline args, it's up to the user to decide whether to turn it on or not

Comment (xen0n) [reply clang/lib/Driver/ToolChains/FreeBSD.cpp:237]: Ah I missed the `hasArg` part. I have no problem with it then...

**Review by xen0n — APPROVED** [review]

File: clang/lib/Driver/ToolChains/FreeBSD.cpp (hunk 3 above)
Comment (hitmoon) [reply clang/lib/Driver/ToolChains/FreeBSD.cpp:237]: > Ah I missed the `hasArg` part. I have no problem with it then...

Thanks !

**Comment by xen0n** [comment]
Also a `[Clang]` tag could be prepended to the PR title to make it clearer what part gets changed.

**Comment by hitmoon** [comment]
clang tag added

**Comment by xen0n** [comment]
> > Also a `[Clang]` tag could be prepended to the PR title to make it clearer what part gets changed.
> 
> Not only clang, all of FreeBSD on LoongArch

It's not about *what can get built after the change*, but *what part of LLVM project this change applies to*, which is Clang. Look at the diff -- all touched files reside in `clang/` directory, that's it.

File: clang/lib/Basic/Targets/OSTargets.h [hunk 4]
```diff
@@ -231,6 +231,9 @@ class LLVM_LIBRARY_VISIBILITY FreeBSDTargetInfo : public OSTargetInfo<Target> {
     case llvm::Triple::riscv32:
     case llvm::Triple::riscv64:
       break;
+    case llvm::Triple::loongarch32:
```
Comment (brooksdavis) [inline clang/lib/Basic/Targets/OSTargets.h:234]: I'm of two minds about adding loongarch32 triples. It's harmless and parallels riscv, but the chances of a new 32-bit port being importing into the FreeBSD are close to zero.

**Review by brooksdavis — APPROVED** [review]

**Review by DimitryAndric — APPROVED** [review]
I don't have any experience with LoongArch, but this diff looks fine to me.

**Review by SixWeining — APPROVED** [review]
LGTM

File: clang/lib/Basic/Targets/OSTargets.h (hunk 4 above)
Comment (hitmoon) [reply clang/lib/Basic/Targets/OSTargets.h:234]: @brooksdavis Thanks!
after have reading [Future of 32-bit platform support in FreeBSD](https://lists.freebsd.org/archives/freebsd-announce/2024-February/000117.html) , I think It's reasonable to exclude 32 bit from LoongArch support. 
I also hope to hear some comments/suggestions from @emaste

Comment (brad0) [reply clang/lib/Basic/Targets/OSTargets.h:234]: @brooksdavis 32-bit RISC-V support was removed for FreeBSD. I just noticed this addition and think it should be removed.

**Comment by alexrp** [comment]
A bit late to the party here, but where is development of FreeBSD for LoongArch taking place? I don't see any traces of it in upstream yet.

**Comment by xen0n** [comment]
> A bit late to the party here, but where is development of FreeBSD for LoongArch taking place? I don't see any traces of it in upstream yet.

Now there is -- they have just posted https://reviews.freebsd.org/D59889 and the entire initial enablement patchset. Seems it's working on QEMU which is a good starting point.

**Pull request merged**